archunit/src/jdk9test/java/com/tngtech/archunit/core/importer/TestLambdaMethodCalls.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright 2019 TNG Technology Consulting GmbH
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.tngtech.archunit.core.importer;
+
+import java.util.List;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+import org.junit.Test;
+
+import com.tngtech.archunit.core.domain.JavaClass;
+import com.tngtech.archunit.core.domain.JavaClasses;
+import com.tngtech.archunit.core.domain.JavaMethod;
+import com.tngtech.archunit.core.importer.ClassFileImporter;
+import com.tngtech.archunit.lang.ArchCondition;
+import com.tngtech.archunit.lang.ArchRule;
+import com.tngtech.archunit.lang.ConditionEvents;
+import com.tngtech.archunit.lang.SimpleConditionEvent;
+import com.tngtech.archunit.lang.syntax.ArchRuleDefinition;
+
+public class TestLambdaMethodCalls {
+  @Test
+  public void testName() throws Exception {
+    JavaClasses importedClasses = new ClassFileImporter()
+        .withImportOption(location -> location.contains("TestLambdaMethodCalls.DummyClassWithLambdas"))
+        .importPackages("com.tngtech.archunit.core.importer");
+
+    ArchRule rule = ArchRuleDefinition.classes().should(haveOnlyNecessaryMethods);
+    rule.check(importedClasses);
+  }
+
+  ArchCondition<JavaClass> haveOnlyNecessaryMethods = new ArchCondition<JavaClass>(
+      "Classes must only have methods that are called at least once") {
+    @Override
+    public void check(final JavaClass item, final ConditionEvents events) {
+      List<JavaMethod> collect = item.getMethods().stream().collect(Collectors.toList());
+
+      for (JavaMethod method : collect) {
```

Comment:
You could directly use `item.getMethods()` and remove `collect`.

Reply:
To extract the essence of this test more clearly, we could even simplify your example to `ArchRuleDefinition.methods().should(beCalled)` using
```java
    ArchCondition<JavaMethod> beCalled = new ArchCondition<JavaMethod>("be called") {
        @Override
        public void check(JavaMethod method, ConditionEvents events) {
            boolean satisfied = !method.getCallsOfSelf().isEmpty();
            String message = String.format("%s %s called", method.getDescription(), satisfied ? "is" : "is not");
            events.add(new SimpleConditionEvent(method, satisfied, message));
        }
    };
```

Reply:
Your example is a nice integration test, but at the end, we probably want to have a unit test (similar to `ClassFileImporterTest`'s `methods_know_callers`) that specifically tests the calls via method references or lambdas.